### ISSUE TITLE
pending actions?

### ISSUE THREAD

**Issue by Antonimo** [lazychaser/laravel-nestedset#42]
What is the idea behind deferred, single action anyway? If I call appendTo I want queries to DB happening.

**Comment by lazychaser** [comment]
The idea is to indicate that when you move your node the data is also saved, that default `save` functionality is behind. So you can do following:

```php
$node->fill($data)->appendTo($otherNode)->save();
```

**Comment by Antonimo** [comment]
between appendTo and save the model is unchanged (not dirty).
in your example, fill does make the model dirty.
Also save acts on the table, not the model only, so the idea of defaultness is gone, as I see it.
I just cant see where its useful =)

**Comment by lazychaser** [comment]
Usefullness in clarity. You still can use model events, for example. Appending node WILL make it dirty (unless it's not actually moved).

> Also save acts on the table, not the model only, so the idea of defaultness is gone, as I see it.
What do you mean?

**Comment by Antonimo** [comment]
appendTo doesn't make it dirty (parent_id is set only on callPendingAction);
I like clarity, and its missing in my opinion. What I mean is responsibility, by example:
```
$model->save(); // model level action, should touch one row
$node->appendTo() // table level action [the nestedset package]
```

**Comment by lazychaser** [comment]
You're right about dirtiness, I looked throught the code and noticed that model will not become dirty if it's parent is not changed.

I think I've started to understand your point. What are your suggestions?

**Comment by lazychaser** [comment]
I think your only concern is that `appendTo` doesn't imply `save`. You can just use this:

```php
$model->fill(compact('parent_id'))->save();
```

This will do everything for you.

Or this:

```php
$parent->append($node);
```

**Comment by Antonimo** [comment]
From my first post: in my opinion, appendTo should act on DB, not wait for save =)

**Comment by lazychaser** [comment]
This will disallow direct change of parent_id and thus dissalow mass asignment of this property which is one of the crucial features. I wanted to integrate nested sets with ELoquent as much as possible, allow users to use `save` as before. Mass asignment, `getAncestors`, `getDescendants` and `toTree` are the most you need.

**Comment by Antonimo** [comment]
Maybe do this:
```
$node->parent_id = 42;
echo $node->parent_id; // 42
```
Set the new parent_id, it will be dirty, and on save() the node is moved as usual...

**Comment by merlinus1** [comment]
Maybe it would be useful to modify Node::setParentIdAttribute method s follows:

```php
public function setParentIdAttribute($value)
    {
        if ($this->getAttribute(static::PARENT_ID) != $value)
        {
            if ($value)
            {
                $this->appendTo($this->newQuery()->findOrFail($value));
            }
            else
            {
                $this->makeRoot();
            }
            $this->attributes[static::PARENT_ID] = $value;
        }
    }
```

it will make $model->parent_id dirty and $model->callPendingAction() will be fired on save.

**Comment by lazychaser** [comment]
@merlinus1 this method is always called and model gets dirty when `actionAppendTo` is performed before actual insert/update is performed.

**Comment by merlinus1** [comment]
It's not get called if parent_id is modified directly and it's only attribute modified.
e.g.:
$model->parent_id = 42;
$model->save();
echo $model->parent_id; //NULL

**Comment by lazychaser** [comment]
@merlinus1 when you set `parent_id` directly, mutator is executed which sets pending action, which is called on `saving` event.

If I'm missing something, please provide failing test.

**Comment by merlinus1** [comment]
Failing test to show that model does not become dirty:

```php
public function testModelIsDirtyAfterDirectAssigment()
{
    $category = Category::create(['name' => 'test']);
    $category->parent_id = 1;
    $this->assertTrue($category->isDirty());
}
```

I think, setting model dirty is useful, because other code may relay on it.
For example see: https://github.com/dimsav/laravel-translatable/blob/master/src/Translatable/Translatable.php#L138

If model is not dirty parent::save does not get called and $model->callPendingAction() never gets caled.

**Comment by lazychaser** [comment]
@merlinus1 so the problem is that `parent_id` is not dirty unless Eloquent's original `save` is hit.

**Comment by samdking** [comment]
Is this still an ongoing discussion? I've come across the above problem recently using v2.4 of this package. I had some previous code that relied on parent_id being dirty, which broke after I swapped in this package.

Apologies, what was the reason for this pull request being closed? https://github.com/lazychaser/laravel-nestedset/pull/60

**Comment by merlinus1** [comment]
@lazychaser is right. Model becomes dirty only if original save handler is hit.

**Comment by samdking** [comment]
But this a problem, right? If I set the parent of a node, whether via `$node->parent_id = $parent->id`, or `$node->parent()->associate($parent)`, or other means, I'd expect inspection of the parent to be that new value (as Antonimo mentioned: https://github.com/lazychaser/laravel-nestedset/issues/42#issuecomment-87223836). On save, this value should persist and the relevant nested set values updated accordingly.

**Comment by lazychaser** [comment]
@samdking you're absolutely right and I'l look into this problem very soon.

**Comment by lazychaser** [comment]
@samdking just noticed that you're using 2.4. It's no longer maintained. I will fix this problem in upcoming patch in v3.1.1.

**Comment by samdking** [comment]
Nice one, thanks!
